### PR TITLE
Document new CPU scheduling flags

### LINKED ISSUE #583: Document support for --cpu-rt-period and --cpu-rt-runtime
### Problem description

Document support for --cpu-rt-period and --cpu-rt-runtime

See https://github.com/docker/docker/pull/23430#pullrequestreview-5960208

### Project version(s) affected

Engine 1.13

cc/ @thaJeztah @erikstmartin

### PR DESCRIPTION
**Pull request by mstanleyjones**
### Describe the proposed changes

Documented new CPU-related flags and tried to pull them together into a coherent topic

### Unreleased project version

Engine 1.13

### Related issue

Fixes #583

### Related issue or PR in another project

https://github.com/docker/docker/pull/23430#pullrequestreview-5960208

### Please take a look

@justincormack @thaJeztah @erikstmartin

### CODE REVIEW

**Comment by mstanleyjones** [comment]
PTAL @cpuguy83

File: engine/admin/container_priority.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+keywords:
+- priority, real time, nice, resources, cpu
+title: Configure a container's CPU scheduling
+---
+
+Containers run as isolated processes on the Docker daemon host. Normally a
+container gets access to CPU resources proportional to all other running
+processes which need CPU resources. If your host machine and Docker daemon are
+configured correctly, you can schedule Linux container tasks to have higher
+priority access to CPU resources. If you need to, you can even schedule
```
Comment (cpuguy83) [inline engine/admin/container_priority.md]: The options Docker provides are really more about being able to configure the resources that the RT scheduler should allot for it, but does not set the scheduler itself.

**Comment by mstanleyjones** [comment]
@mlaventure PTAL, Brian nominated you. :)

**Comment by mstanleyjones** [comment]
I realized that this needs to wait for #334 is merged into master, master is merged into vnext-engine, and this stuff needs to go into that topic instead of this new topic I've created. Just a note for me, no action for reviewers.

**Comment by mstanleyjones** [comment]
OK, this work has now been refactored on top of #334. No extra review is needed because no facts were changed.

**Comment by vincentwoo** [comment]
I think this documentation should include more information about realtime scheduling and links to places to find out how to enable it for various distros. Docker introduced a pretty useful new feature without a great plan for users on how to actually get it running in prod. For instance, if you want to use `--cpus`, do you need to cap-add sys_nice to all your containers?

**Comment by mstanleyjones** [comment]
@cpuguy83 do you have any info to help answer @vincentwoo's question?

**Comment by vincentwoo** [comment]
There's a lot more information in the description/comments of the originating PR (https://github.com/docker/docker/pull/23430) but no information in the documentation about how much of this is still needed, nor what their base assumptions are. I think it likely that *many* users (including myself) have no idea what "real time scheduling" is.

**Comment by mstanleyjones** [comment]
I'm not sure how much about real-time scheduling as a computer science concept or a Linux kernel concept that Docker should be covering. I think in general there is an idea that "If you are not sure whether you need RT, you don't". I think we already link to the kernel references, and I'm not sure how far we should be going. I'll wait for @cpuguy83 or others to weigh in.

**Comment by vincentwoo** [comment]
I completely disagree. The docs present the new cpu options as *just-so*. As in, simply pass these flags to your containers and they will be throttled accordingly. `--cpus=1.5` seems like a great thing, and very easy for anyone to understand! The layout of the documentation makes it seem like you can just turn it on for any container because it appears under "Configure the default scheduler".

Then, the docs go on to describe a bunch of weird options, like realtime, or sys_nice, with no mention of *why* you might want to turn on those options. Is the reality that you actually need all of that stuff to use `--cpus`? Because from reading the other PR it seems like the answer is definitely, yes.

**Comment by vincentwoo** [comment]
See also https://github.com/docker/docker/pull/27958, which introduced `--cpus` on top of the previous PR. What's also very weird is all this instruction around "If you use Docker 1.13 or higher, use --cpus instead". **ALL** of these flags were introduced in 1.13, no?

**Comment by cpuguy83** [comment]
@vincentwoo I think you are mixing things up here.
`--cpus` configures CFS, and it works just as is.

If someone is using realtime workloads `--cpus` is not for them. The docs for `--cpus` is specifically under the CFS scheduler options.

These options seem well separated in the docs.

**Comment by mstanleyjones** [comment]
Maybe it would help to see this content in context: https://docs.docker.com/engine/admin/resource_constraints/#/cpu

**Comment by vincentwoo** [comment]
Oh jesus I finally understand. I apologize for my inability to understand the docs - did not realize that different PRs were talking about runtime vs cfs periods, ugh. apologies to all involved

**Pull request merged**